jumanji/environments/routing/connector/generation_methods.py

```diff
@@ -0,0 +1,437 @@
+                (action_tuple == jnp.array([1, 0])).all(axis=0),
+                (action_tuple == jnp.array([0, -1])).all(axis=0),
+                (action_tuple == jnp.array([0, 1])).all(axis=0),
+                (action_tuple == jnp.array([0, 0])).all(axis=0),
+            ]
+        )
+        actions = jnp.sum(actions * action_multiplier, axis=0)
+        return actions
+
+    def _adjacent_cells(self, cell: int) -> chex.Array:
+        """Returns chex.Array of adjacent cells to the input.
+
+        Given a cell, return a chex.Array of size 4 with the flat indices of
+        adjacent cells. Padded with -1's if less than 4 adjacent cells (if on the edge of the grid).
+
+        Args:
+            cell: the flat index of the cell to find adjacent cells of.
+
+        Returns:
+            A chex.Array of size 4 with the flat indices of adjacent cells
+            (padded with -1's if less than 4 adjacent cells).
+        """
+        available_moves = jnp.full(4, cell)
+        direction_operations = jnp.array([-1 * self.rows, self.rows, -1, 1])
+        # Create a mask to check 0 <= index < total size
+        cells_to_check = available_moves + direction_operations
+        is_id_in_grid = cells_to_check < self.rows * self.cols
+        is_id_positive = 0 <= cells_to_check
+        mask = is_id_positive & is_id_in_grid
+
+        # Ensure adjacent cells doesn't involve going off the grid
+        unflatten_available = jnp.divmod(cells_to_check, self.cols)
+        unflatten_current = jnp.divmod(cell, self.cols)
+        is_same_row = unflatten_available[0] == unflatten_current[0]
+        is_same_col = unflatten_available[1] == unflatten_current[1]
+        row_col_mask = is_same_row | is_same_col
+        # Combine the two masks
+        mask = mask & row_col_mask
+        return jnp.where(mask == 0, -1, cells_to_check)
+
+    def _available_cells(self, grid: chex.Array, cell: chex.Array) -> chex.Array:
+        """Returns list of cells that can be stepped into from the input cell's position.
+
+        Given a cell and the grid of the board, see which adjacent cells are available to move to
+        (i.e. are currently unoccupied) to avoid stepping over exisitng wires.
+
+        Args:
+            grid: the current layout of the board i.e. current grid.
+            cell: the flat index of the cell to find adjacent cells of.
+
+        Returns:
+            A chex.Array of size 4 with the flat indices of adjacent cells.
+        """
+        adjacent_cells = self._adjacent_cells(cell)
+        # Get the wire id of the current cell
+        value = grid[jnp.divmod(cell, self.cols)]
+        wire_id = (value - 1) // 3
+
+        _, available_cells_mask = jax.lax.scan(self._is_cell_free, grid, adjacent_cells)
+        # Also want to check if the cell is touching itself more than once
+        _, touching_cells_mask = jax.lax.scan(
+            self._is_cell_doubling_back, (grid, wire_id), adjacent_cells
+        )
```

Comment:
This can be vmap-ed to parallelise the computation:
```suggestion
        touching_cells_mask = jax.vmap(self._is_cell_doubling_back, in_axes=(None, None, 0))(grid, wire_id, adjacent_cells)
```

Reply:
Resolved. Had to change the signature of one of the other functions but all good now